File: zig/ui/glfw/vertex_buffer.zig
const std = @import("std");
const gl = @cImport(@cInclude("GL/glew.h"));


pub const VertexBuffer = struct {
    rendererID: u32 = 0,

    pub fn destroy(self: *VertexBuffer) void {
        gl.glDeleteBuffers().?(1, &self.rendererID);

    }

    pub fn bind(self: VertexBuffer) void {
        gl.glBindBuffer().?(gl.GL_ARRAY_BUFFER, self.rendererID);

    }

    pub fn unbind(self: VertexBuffer) void {
        _ = self;
        gl.glBindBuffer().?(gl.GL_ARRAY_BUFFER, 0);
    }
};


pub fn vertexBuffer(data: anyopaque, size: u32) VertexBuffer {
    var vb: VertexBuffer = {};
    gl.glGenBuffers().?(1, &vb.rendererID);
    gl.glBindBuffer().?(gl.GL_ARRAY_BUFFER, vb.rendererID);
    gl.glBufferData().?(gl.GL_ARRAY_BUFFER, data, size, gl.GL_STATIC_DRAW);
    return vb;
}

File: c++/ui/opengl/vertex_buffer.cpp
#include "vertex_buffer.hpp"

VertexBuffer::VertexBuffer(void *data, u32 size) {
    glGenBuffers(1, &this->rendererID);
    glBindBuffer(GL_ARRAY_BUFFER, this->rendererID);
    glBufferData(GL_ARRAY_BUFFER, size, data, GL_STATIC_DRAW);
}

VertexBuffer::~VertexBuffer() {
    glDeleteBuffers(1, &this->rendererID);
}

void VertexBuffer::bind() {
    glBindBuffer(GL_ARRAY_BUFFER, this->rendererID);

}
void VertexBuffer::unbind() {
    glBindBuffer(GL_ARRAY_BUFFER, 0);
}

